packages/wdio-shared-store-service/src/client.ts

```diff
@@ -29,40 +37,48 @@ export const getValue = async (key: string): Promise<string | number | boolean |
  * @param {*}       value `store[key]` value (plain object)
  */
 export const setValue = async (key: string, value: JsonCompatible | JsonPrimitive) => {
-    /**
-     * if someone calls `setValue` in `onPrepare` we don't have a base url
-     * set as the launcher is called after user hooks. In this case we need
-     * to wait until it is set and flush all messages.
-     */
-    if (baseUrl) {
+    const setPromise = baseUrlPromise.then((baseUrl) => {
         return got.post(`${baseUrl}/set`, { json: { key, value } }).catch(errHandler)
-    }
+    })
 
-    log.info('Shared store server not yet started, collecting value')
-    pendingValues.set(key, value)
+    return isBaseUrlReady ? setPromise : Promise.resolve()
+}
 
-    if (waitTimeout) {
-        return
-    }
+/**
+ *
+ * @param {string}  key
+ * @param {*}       value
+ */
+export const setResourcePool = async (key: string, value: JsonArray) => {
+    const setPromise = baseUrlPromise.then((baseUrl) => {
+        return got.post(`${baseUrl}/pool/set`, { json: { key, value } }).catch(errHandler)
+    })
 
-    log.info('Check shared store server to start')
-    waitTimeout = setInterval(async () => {
-        if (!baseUrl) {
-            return
-        }
+    return isBaseUrlReady ? setPromise : Promise.resolve()
+}
 
-        log.info(`Shared store server started, flushing ${pendingValues.size} values`)
-        clearInterval(waitTimeout)
-        await Promise.all([...pendingValues.entries()].map(async ([key, value]) => {
-            await got.post(`${baseUrl}/set`, { json: { key, value } }).catch(errHandler)
-            pendingValues.delete(key)
-        })).then(
-            () => log.info('All pending values were successfully stored'),
-            (err) => log.error(`Failed to store all values: ${err.stack}`)
-        )
-    }, WAIT_INTERVAL)
+/**
+ *
+ * @param {string}  key
+ * @param {*}       value
+ */
+export const getValueFromPool = async (key: string, options: GetValueOptions) => {
+    const baseUrl = await baseUrlPromise
+    const res = await got.get(`${baseUrl}/pool/get/${key}${options?.timeout ? `?timeout=${options.timeout}` : '' }`, { responseType: 'json' }).catch(errHandler)
+    return res?.body ? (res.body as JsonObject).value : undefined
+}
+
+/**
+ *
+ * @param {string}  key
+ * @param {*}       value
+ */
+export const addValueToPool = async (key: string, value: JsonPrimitive | JsonCompatible) => {
+    const baseUrl = await baseUrlPromise
+    const res = await got.post(`${baseUrl}/pool/add`, { json: { key, value }, responseType: 'json' }).catch(errHandler)
+    return res?.body ? (res.body as JsonObject).value : undefined
 }
 
-const errHandler = (err: Response<Error>) => {
-    log.warn(err.statusCode, err.statusMessage, err.url, err.body)
+const errHandler = (err: RequestError) => {
+    throw new Error(`${err.response?.body || 'Shared store server threw an error'}`)
```

Comment:
@christian-bromann please validate this

before: the polka server would crash on error throw and all wdio process crashed suddenly
now: the polka server returns an error, this client throws it to wdio, the message is logged, the worker crashes, but wdio process continues on to the next test

i think this was the desired behaviour, not to immediately crash all running suites

Reply:
Sounds good to me 👍